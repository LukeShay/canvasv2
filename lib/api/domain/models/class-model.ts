/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable global-require */
import { Model, RelationMappings, JSONSchema } from 'objection';
import Constants from '../constants';

export class ClassModel extends Model {
  static tableName = Constants.CLASSES_TABLE;

  static idColumn = 'id';

  static get jsonSchema(): JSONSchema {
    return {
      properties: {
        adminId: { type: 'string' },
        building: { type: ['string', 'null'] },
        code: { type: 'string' },
        description: { type: ['string', 'null'] },
        name: { type: 'string' },
        room: { type: ['string', 'null'] },
      },
      required: ['adminId', 'name', 'code'],
      type: 'object',
    };
  }

  static get relationMappings(): RelationMappings {
    // ts-expect-error
    const { UserModel } = require('./user-model');
    // ts-expect-error
    const { AssistantModel } = require('./assistant-model');

    return {
      admin: {
        join: {
          from: `${ClassModel.tableName}.adminId`,
          to: `${UserModel.tableName}.${UserModel.idColumn}`,
        },
        modelClass: UserModel,
        relation: Model.BelongsToOneRelation,
      },
      assistants: {
        join: {
          from: `${ClassModel.tableName}.${ClassModel.idColumn}`,
          to: `${AssistantModel.tableName}.classId`,
        },
        modelClass: AssistantModel,
        relation: Model.HasManyRelation,
      },
    };
  }
}
