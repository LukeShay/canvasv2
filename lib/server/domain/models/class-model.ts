/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable global-require */
import { Model, RelationMappings, JSONSchema } from 'objection';
import Constants from '../constants';

export class ClassModel extends Model {
  static tableName = Constants.CLASSES_TABLE;

  static idColumn = 'id';

  static get jsonSchema(): JSONSchema {
    return {
      properties: {
        adminId: { type: 'string' },
        building: { type: ['string', 'null'] },
        code: { type: 'string' },
        description: { type: ['string', 'null'] },
        name: { type: 'string' },
        room: { type: ['string', 'null'] },
      },
      required: ['adminId', 'name', 'code'],
      type: 'object',
    };
  }

  static get relationMappings(): RelationMappings {
    const { UserModel } = require('./user-model');
    const { AssistantModel } = require('./assistant-model');
    const { StudentModel } = require('./student-model');

    return {
      admin: {
        join: {
          from: `${ClassModel.tableName}.adminId`,
          to: `${UserModel.tableName}.${UserModel.idColumn}`,
        },
        modelClass: UserModel,
        relation: Model.BelongsToOneRelation,
      },
      assistants: {
        join: {
          from: `${ClassModel.tableName}.${ClassModel.idColumn}`,
          through: {
            from: `${AssistantModel.tableName}.classId`,
            to: `${AssistantModel.tableName}.assistantId`,
          },
          to: `${UserModel.tableName}.${UserModel.tableName}`,
        },
        modelClass: UserModel,
        relation: Model.ManyToManyRelation,
      },
      students: {
        join: {
          from: `${ClassModel.tableName}.${ClassModel.idColumn}`,
          through: {
            from: `${StudentModel.tableName}.classId`,
            to: `${StudentModel.tableName}.studentId`,
          },
          to: `${UserModel.tableName}.${UserModel.tableName}`,
        },
        modelClass: UserModel,
        relation: Model.ManyToManyRelation,
      },
    };
  }
}
