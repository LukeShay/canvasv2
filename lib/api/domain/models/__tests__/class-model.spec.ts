import { ClassModel } from '../class-model';

describe('ClassModel', () => {
  it("should have idColum 'id'", () => {
    expect(ClassModel.idColumn).toEqual('id');
  });

  it('should should have jsonSchema that matches snapshot', () => {
    expect(ClassModel.jsonSchema).toMatchSnapshot();
  });

  it('should should have relationMappings that matches snapshot', () => {
    expect(ClassModel.relationMappings).toMatchSnapshot();
  });
});
