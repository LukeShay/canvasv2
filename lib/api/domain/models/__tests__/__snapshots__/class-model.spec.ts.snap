// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClassModel should should have jsonSchema that matches snapshot 1`] = `
Object {
  "properties": Object {
    "adminId": Object {
      "type": "string",
    },
    "building": Object {
      "type": Array [
        "string",
        "null",
      ],
    },
    "code": Object {
      "type": "string",
    },
    "description": Object {
      "type": Array [
        "string",
        "null",
      ],
    },
    "name": Object {
      "type": "string",
    },
    "room": Object {
      "type": Array [
        "string",
        "null",
      ],
    },
  },
  "required": Array [
    "adminId",
    "name",
    "code",
  ],
  "type": "object",
}
`;

exports[`ClassModel should should have relationMappings that matches snapshot 1`] = `
Object {
  "admin": Object {
    "join": Object {
      "from": "Classes.adminId",
      "to": "Users.id",
    },
    "modelClass": [Function],
    "relation": [Function],
  },
  "assistants": Object {
    "join": Object {
      "from": "Classes.id",
      "to": "Assistants.classId",
    },
    "modelClass": [Function],
    "relation": [Function],
  },
}
`;
